sqlmesh/core/model/kind.py

```diff
@@ -81,18 +92,80 @@ def model_kind_name(self) -> ModelKindName:
 class ModelKind(PydanticModel, ModelKindMixin):
     name: ModelKindName
 
-    def to_expression(self, **kwargs: t.Any) -> d.ModelKind:
-        return d.ModelKind(this=self.name.value.upper(), **kwargs)
+    @classmethod
+    def field_validator(cls) -> classmethod:
+        def _model_kind_validator(v: t.Any) -> ModelKind:
+            if isinstance(v, ModelKind):
+                return v
+
+            if isinstance(v, d.ModelKind):
+                name = v.this
+                props = {prop.name: prop.args.get("value") for prop in v.expressions}
+                klass: t.Type[ModelKind] = ModelKind
+                if name == ModelKindName.INCREMENTAL_BY_TIME_RANGE:
+                    klass = IncrementalByTimeRangeKind
+                elif name == ModelKindName.INCREMENTAL_BY_UNIQUE_KEY:
+                    klass = IncrementalByUniqueKeyKind
+                elif name == ModelKindName.SEED:
+                    klass = SeedKind
+                else:
+                    props["name"] = ModelKindName(name)
+                return klass(**props)
+
+            if isinstance(v, dict):
+                if v.get("name") == ModelKindName.INCREMENTAL_BY_TIME_RANGE:
+                    klass = IncrementalByTimeRangeKind
+                elif v.get("name") == ModelKindName.INCREMENTAL_BY_UNIQUE_KEY:
+                    klass = IncrementalByUniqueKeyKind
+                elif v.get("name") == ModelKindName.SEED:
+                    klass = SeedKind
+                else:
+                    klass = ModelKind
+                return klass(**v)
+
+            name = (v.name if isinstance(v, exp.Expression) else str(v)).upper()
+
+            try:
+                return ModelKind(name=ModelKindName(name))
+            except ValueError:
+                raise ConfigError(f"Invalid model kind '{name}'")
+
+        return validator("kind", pre=True, allow_reuse=True)(_model_kind_validator)
```

Comment:
Why are we returning the `validator` itself? Can't we just do something like:
```
_kind_validator = validator("kind", pre=True, allow_reuse=True)(_model_kind_validator)
```

Reply:
We already follow a similar pattern here: https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/model/common.py#L47

Reply:
Ah, I see now that you moved away from it. Why?

Reply:
Well I kept the pattern in that we are still sharing these validators. I just now have them inside the class that they are used to validate again so it is easier to reference the validator you need. At one point I was adding more validators and this felt like a cleaner way to organize them.

Reply:
That's ok. Do we need to create a new `validator` every time, though, or can can it just be a class member?